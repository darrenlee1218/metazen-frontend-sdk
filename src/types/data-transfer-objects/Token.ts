interface TokenDisplay {
  /**
   * specific color for display in some places
   */
  colorCode: string;

  /**
   * URL of token icon
   */
  pngIconUrl: string;

  /**
   * URL of token icon
   */
  svgIconUrl: string;

  /**
   * NFT collection icon
   */
  nftCollectionImageUrl?: string;

  /**
   * background image for the NFT collection
   */
  nftCollectionBackgroundImageUrl?: string;
}

interface TokenChain {
  /**
   * unique chain ID
   */
  chainID: string;

  /**
   * name of the chain
   */
  chainName: string;

  /**
   * whether it's a testnet/mainnet chain
   */
  type: 'TESTNET' | 'MAINNET';

  /**
   * minimum block confirmatoin required for a transaction
   */
  minBlockConfirmation: number;

  /**
   * Smart contract URL
   */
  smartContractUrl: string;

  /**
   * Address URL (for reviewing wallet info)
   */
  addressUrl: string;

  /**
   * tx URL (for reviewing address history)
   */
  txUrl: string;
}

interface TokenNetwork {
  /**
   * network name for chainalysis usage
   */
  chainalysisNetwork: string;

  /**
   * color display for the network
   */
  networkColor: string;

  /**
   * URL of network icon
   */
  pngNetworkIconUrl: string;
}

// TODO: set Token['standard] = TokenStandard
export enum TokenStandard {
  NATIVE = 'NATIVE',
  ERC20 = 'ERC-20',
  ERC721 = 'ERC-721',
  ERC1155 = 'ERC-1155',
}

export default interface Token {
  /**
   * unique token ID, from chain ID, ticker and contract address (contract addr only for non-native token)
   */
  key: string;

  /**
   * token name
   */
  displayName: string;

  /**
   * asset ticker
   * WARNING: ticker may be not unique
   */
  ticker: string;

  /**
   * standard of an token (e.g. ERC-1155)
   * always empty for native token
   */
  standard: TokenStandard;

  /**
   * whether it is a native token
   */
  isNative: boolean;

  /**
   * token's Explorer address i.e.polygonscan
   */
  txUrl: string;

  /**
   * token's contract address
   * undefined for native token
   */
  contractAddress: string | undefined;

  /**
   * gas limit of token for estimate gas fee
   */
  gasLimit: string;

  /**
   * decimal place for token amount
   */
  decimal?: number;

  /**
   * network related stuff
   */
  network: TokenNetwork;

  /**
   * decimal places for custody
   * defined for all ERC-20 / Native tokens
   */
  custodyPrecision?: number;

  /**
   * decimal places for display
   * defined for all ERC-20 / Native tokens
   */
  displayPrecision?: number;
  /**
   * display-related properties of an token
   */
  display: TokenDisplay;

  /**
   * chain info
   */
  chain: TokenChain;
}
