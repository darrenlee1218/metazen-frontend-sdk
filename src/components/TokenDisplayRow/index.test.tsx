import React from 'react';
import { render, fireEvent, screen, cleanup } from '@testing-library/react';
import { ComponentProps } from 'react';
import { TokenDisplayRow } from '.';
import { TestProviders } from '@utils/testing/TestProviders';
import Token, { TokenStandard } from '@gryfyn-types/data-transfer-objects/Token';

afterEach(cleanup);

const defaultProps: ComponentProps<typeof TokenDisplayRow> = {
  token: {
    isNative: false,
    displayName: 'Polygon (Mainnet)',
    network: {
      pngNetworkIconUrl: 'network.png',
    },
    display: {
      pngIconUrl: 'icon.png',
      svgIconUrl: 'icon.svg',
    },
    ticker: 'MATIC',
    displayPrecision: 4,
    custodyPrecision: 8,
    decimal: 18,
    standard: TokenStandard.ERC20,
  } as unknown as Token,
  price: '1',
  balanceInAssetDecimals: '10000000000000000000',
};

it('calls the callback after a click event', () => {
  const callback = jest.fn();
  const props = { ...defaultProps, onClick: callback };
  render(<TokenDisplayRow {...props} />, { wrapper: TestProviders });
  fireEvent.click(screen.getByText(defaultProps.token.displayName));
  expect(callback).toHaveBeenCalled();
});

it('does not render prices if price / amountUsdValue is falsey', () => {
  const props = { ...defaultProps, price: '', amountUsdValue: undefined };
  render(<TokenDisplayRow {...props} />, { wrapper: TestProviders });
  expect(screen.queryByTestId('prices-row')).toBe(null);
});

it('matches snapshot', () => {
  const { container } = render(<TokenDisplayRow {...defaultProps} />, { wrapper: TestProviders });
  expect(container).toMatchSnapshot();
});
