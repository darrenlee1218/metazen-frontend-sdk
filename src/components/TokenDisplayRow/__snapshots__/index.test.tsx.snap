// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiCardActionArea-root MuiBox-root css-29ttzx-MuiButtonBase-root-MuiCardActionArea-root"
    tabindex="0"
    type="button"
  >
    <div
      class="MuiBox-root css-16bryr0"
    >
      <div
        class="MuiBox-root css-1u6ef5p"
      >
        <div
          class="MuiBox-root css-iiu66e"
        />
      </div>
    </div>
    <div
      class="MuiBox-root css-zg1vud"
    >
      <div
        class="MuiBox-root css-qf05oz"
        data-testid="name-balance-row"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2 css-m3q21w-MuiTypography-root"
        >
          Polygon (Mainnet)
        </h2>
        <h2
          class="MuiTypography-root MuiTypography-h2 css-g1bq1m-MuiTypography-root"
          data-testid="amount-display"
        >
          10 MATIC
        </h2>
      </div>
      <div
        class="MuiBox-root css-qf05oz"
        data-testid="prices-row"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 css-5h7vtp-MuiTypography-root"
          data-testid="currency-display"
        >
          $1.0000
        </h4>
        <h4
          class="MuiTypography-root MuiTypography-h4 css-5h7vtp-MuiTypography-root"
          data-testid="currency-display"
        >
          $10.00
        </h4>
      </div>
    </div>
    <span
      class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
    />
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
